Expand and re-organize InfluxDB Enterprise authn and authz docs (#3330)
influxdata/docs-v2: content/enterprise_influxdb/v1.9/administration/configure/security/authentication.md

    name: Configure authentication
weight: 10
---

{{% enterprise-warning-authn-b4-authz %}}

## Enable authentication

Authentication is disabled by default in InfluxDB and InfluxDB Enterprise.
After [installing the data nodes](/enterprise_influxdb/v1.9/introduction/install-and-deploy/installation/data_node_installation/)),
we recommend enabling authentication to control access to your cluster.

To enable authentication in a cluster, do the following:

1. Set `auth-enabled` to `true` in the `[http]` section of the config files 
   for all meta **and** data nodes:
   ```toml
   [http]
     # ...
     auth-enabled = true
   ```
1. Next, create an admin user (if you haven't already).
   Using the [`influx` CLI](/enterprise_influxdb/v1.9/tools/influx-cli/),
   run the following command:
   ```
   CREATE USER admin WITH PASSWORD 'mypassword' WITH ALL PRIVILEGES
   ```
1. Restart InfluxDB Enterprise.
   Once restarted, InfluxDB Enterprise checks user credentials on every request
   and only processes requests with valid credentials.

## Configure authentication using JWT tokens

For a more secure alternative to using passwords,
include JWT tokens with requests to the InfluxDB API.

> This is currently only possible through the [InfluxDB HTTP API](/enterprise_influxdb/v1.9/tools/api/).

Author: ```suggestion
This is only available through the [InfluxDB HTTP API](/enterprise_influxdb/v1.9/tools/api/).
```

   InfluxDB Enterprise uses the shared secret to encode the JWT signature.
   By default, `shared-secret` is set to an empty string, in which case no JWT authentication takes place.
   <!-- TODO: meta, data, or both? -->
   Add a custom shared secret in your [InfluxDB configuration file](/enterprise_influxdb/v1.9/administration/configure/config-data-nodes/#shared-secret--).
   The longer the secret string, the more secure it is:

   ```toml
   [http]
   shared-secret = "my super secret pass phrase"
   ```

   Alternatively, to avoid keeping your secret phrase as plain text in your InfluxDB configuration file,
   set the value with the `INFLUXDB_HTTP_SHARED_SECRET` environment variable.

2. **Generate your JWT token**.

   Use an authentication service to generate a secure token
   using your InfluxDB username, an expiration time, and your shared secret.
   There are online tools, such as [https://jwt.io/](https://jwt.io/), that will do this for you.

   The payload (or claims) of the token must be in the following format:

   ```json
   {
       "username": "myUserName",
       "exp": 1516239022
   }
   ```

   - **username** - The name of your InfluxDB user.

>    - **exp** - The expiration time of the token in UNIX epoch time.

Author: ```suggestion
   - **exp** - The expiration time of the token in UNIX [epoch time](/enterprise_influxdb/v1.9/query_language/explore-data/#epoch_time).
```

To enable authentication in a cluster, do the following:

1. Set `auth-enabled` to `true` in the `[http]` section of the config files 
   for all meta **and** data nodes:
   ```toml
   [http]
     # ...
     auth-enabled = true
   ```
1. Next, create an admin user (if you haven't already).
   Using the [`influx` CLI](/enterprise_influxdb/v1.9/tools/influx-cli/),
   run the following command:
   ```
   CREATE USER admin WITH PASSWORD 'mypassword' WITH ALL PRIVILEGES
   ```
1. Restart InfluxDB Enterprise.
   Once restarted, InfluxDB Enterprise checks user credentials on every request
   and only processes requests with valid credentials.

## Configure authentication using JWT tokens

For a more secure alternative to using passwords,
include JWT tokens with requests to the InfluxDB API.
This is currently only possible through the [InfluxDB HTTP API](/enterprise_influxdb/v1.9/tools/api/).

1. **Add a shared secret in your InfluxDB Enterprise configuration file**.

   InfluxDB Enterprise uses the shared secret to encode the JWT signature.
   By default, `shared-secret` is set to an empty string, in which case no JWT authentication takes place.

>    <!-- TODO: meta, data, or both? -->

Reviewer: still tbd?

   Add a custom shared secret in your [InfluxDB configuration file](/enterprise_influxdb/v1.9/administration/configure/config-data-nodes/#shared-secret--).

>    The longer the secret string, the more secure it is:

Reviewer: ```suggestion
   Longer strings are more secure:
```

1. **Add a shared secret in your InfluxDB Enterprise configuration file**.

   InfluxDB Enterprise uses the shared secret to encode the JWT signature.
   By default, `shared-secret` is set to an empty string, in which case no JWT authentication takes place.
   <!-- TODO: meta, data, or both? -->
   Add a custom shared secret in your [InfluxDB configuration file](/enterprise_influxdb/v1.9/administration/configure/config-data-nodes/#shared-secret--).
   Longer strings are more secure:

   ```toml
   [http]
   shared-secret = "my super secret pass phrase"
   ```

   Alternatively, to avoid keeping your secret phrase as plain text in your InfluxDB configuration file,
   set the value with the `INFLUXDB_HTTP_SHARED_SECRET` environment variable.

2. **Generate your JWT token**.

   Use an authentication service to generate a secure token
   using your InfluxDB username, an expiration time, and your shared secret.
   There are online tools, such as [https://jwt.io/](https://jwt.io/), that will do this for you.

   The payload (or claims) of the token must be in the following format:

   ```json
   {
       "username": "myUserName",
       "exp": 1516239022
   }

>    ```

Author: The payload is the input, and the "generated token" refers to the output, after signing and encoding. <img width="1271" alt="Screen Shot 2022-01-05 at 10 07 58 AM" src="https://user-images.githubusercontent.com/19642016/148249932-adbcbec2-6eb7-4424-8200-d04d2dac38be.png">

Still, line 86 is probably superfluous, since the output format is standard.

   - **username** - The name of your InfluxDB user.
   - **exp** - The expiration time of the token in UNIX [epoch time](/enterprise_influxdb/v1.9/query_language/explore-data/#epoch_time).
     For increased security, keep token expiration periods short.
     For testing, you can manually generate UNIX timestamps using [https://www.unixtimestamp.com/index.php](https://www.unixtimestamp.com/index.php).

   To encode the payload using your shared secret, use a JWT library in your own authentication server or encode by hand at [https://jwt.io/](https://jwt.io/).

   The generated token follows this format: `<header>.<payload>.<signature>`

3. **Include the token in HTTP requests**.

   Include your generated token as part of the `Authorization` header in HTTP requests:

   ```
   Authorization: Bearer <myToken>
   ```
   {{% note %}}
Only unexpired tokens will successfully authenticate.
Be sure your token has not expired.
   {{% /note %}}

#### Example query request with JWT authentication
```bash
curl -G "http://localhost:8086/query?db=demodb" \
  --data-urlencode "q=SHOW DATABASES" \

>   --header "Authorization: Bearer <header>.<payload>.<signature>"

Author: I don't know. Maybe @jstirnaman would have a better idea?

---
title: Configure authentication
description: >
  Enable authentication to require credentials for a cluster.
menu:
  enterprise_influxdb_1_9:
    parent: Configure security
    name: Configure authentication
weight: 10
---

To configure authentication, do one of the following: 

- [Enable authentication](#enable-authentication)

> - [Configure authentication using JWT tokens](#configure-authentication-using-jwt-tokens) ([InfluxDB HTTP API](/enterprise_influxdb/v1.9/tools/api/) only)

Reviewer: ```suggestion
- ([InfluxDB HTTP API](/enterprise_influxdb/v1.9/tools/api/) only)[Configure authentication using JWT tokens](#configure-authentication-using-jwt-tokens) 
```
